packages/react-instantsearch-hooks-server/src/getServerState.tsx

```diff
@@ -27,37 +30,64 @@ export function getServerState(
   };
 
-  ReactDOM.renderToString(
-    <InstantSearchServerContext.Provider value={{ notifyServer }}>
-      {children}
-    </InstantSearchServerContext.Provider>
-  );
-
-  // We wait for the component to mount so that `notifyServer()` is called.
-  return new Promise((resolve) => setTimeout(resolve, 0))
-    .then(() => {
-      // If `notifyServer()` is not called by then, it means that <InstantSearch>
-      // wasn't within the `children`.
-      // We decide to go with a strict behavior in that case; throwing. If users have
-      // some routes that don't mount the <InstantSearch> component, they would need
-      // to try/catch the `getServerState()` call.
-      // If this behavior turns out to be too strict for many users, we can decide
-      // to warn instead of throwing.
-      if (!searchRef.current) {
-        throw new Error(
-          "Unable to retrieve InstantSearch's server state in `getServerState()`. Did you mount the <InstantSearch> component?"
-        );
-      }
-
-      return waitForResults(searchRef.current);
-    })
-    .then(() => {
-      const initialResults = getInitialResults(searchRef.current!.mainIndex);
-
-      return {
-        initialResults,
-      };
+  // eslint-disable-next-line no-shadow
+  function execute(children: ReactNode) {
+    ReactDOM.renderToString(
+      <InstantSearchServerContext.Provider value={{ notifyServer }}>
+        {children}
+      </InstantSearchServerContext.Provider>
+    );
+
+    // We wait for the component to mount so that `notifyServer()` is called.
+    return new Promise((resolve) => setTimeout(resolve, 0))
+      .then(() => {
+        // If `notifyServer()` is not called by then, it means that <InstantSearch>
+        // wasn't within the `children`.
+        // We decide to go with a strict behavior in that case; throwing. If users have
+        // some routes that don't mount the <InstantSearch> component, they would need
+        // to try/catch the `getServerState()` call.
+        // If this behavior turns out to be too strict for many users, we can decide
+        // to warn instead of throwing.
+        if (!searchRef.current) {
+          throw new Error(
+            "Unable to retrieve InstantSearch's server state in `getServerState()`. Did you mount the <InstantSearch> component?"
+          );
+        }
+
+        return waitForResults(searchRef.current);
+      })
+      .then(() => {
+        const initialResults = getInitialResults(searchRef.current!.mainIndex);
+
+        return {
+          initialResults,
+        };
+      });
+  }
+
+  return execute(children).then((serverState) => {
+    let shouldRefetch = false;
+
+    // <DynamicWidgets> requires another query to retrieve the dynamic widgets
+    // to render.
+    walkIndex(searchRef.current!.mainIndex, (index) => {
+      shouldRefetch =
+        shouldRefetch ||
+        index
+          .getWidgets()
+          .some((widget) => widget.$$type === 'ais.dynamicWidgets');
     });
```

Comment:
Do you want to refactor this logic to break the loop early in this PR?

Reply:
I thought the numbers of index widgets likely isn't big enough to make a difference

Reply:
To me the main argument is that it's hard to understand this code, based on what it's doing. But I'm fine addressing this later.

Let's keep this conversation unresolved to keep it visible then.

Reply:
alternative:

```js
    let allWidgets: string[] = [];
    walkIndex(searchRef.current!.mainIndex, (index) => {
      allWidgets = allWidgets.concat(
        index.getWidgets().map((widget) => widget.$$type)
      );
    });
    // <DynamicWidgets> requires another query to retrieve the dynamic widgets
    // to render.
    const shouldRefetch = allWidgets.indexOf('ais.dynamicWidgets') > -1;
```